src/lm/lm_chainer.py

```diff
@@ -115,32 +112,124 @@ def predict(self, state, x):
             param.data[...] = np.random.uniform(-0.1, 0.1, param.data.shape)
+        self.n_layers = n_layers
 
     def __call__(self, state, x):
         if state is None:
-            state = {'c1': None, 'h1': None, 'c2': None, 'h2': None}
-        h0 = self.embed(x)
-        c1, h1 = self.l1(state['c1'], state['h1'], F.dropout(h0))
-        c2, h2 = self.l2(state['c2'], state['h2'], F.dropout(h1))
-        y = self.lo(F.dropout(h2))
-        state = {'c1': c1, 'h1': h1, 'c2': c2, 'h2': h2}
+            state = {'c': [None] * self.n_layers, 'h': [None] * self.n_layers}
+        h = [None] * self.n_layers
+        c = [None] * self.n_layers
+        emb = self.embed(x)
+        c[0], h[0] = self.lstm[0](state['c'][0], state['h'][0], F.dropout(emb))
+        for n in six.moves.range(1, self.n_layers):
+            c[n], h[n] = self.lstm[n](state['c'][n], state['h'][n], F.dropout(h[n - 1]))
+        y = self.lo(F.dropout(h[-1]))
+        state = {'c': c, 'h': h}
         return state, y
 
 
+class BPTTUpdater(training.updaters.StandardUpdater):
+
+    def __init__(self, train_iter, optimizer, device):
+        super(BPTTUpdater, self).__init__(
+            train_iter, optimizer, device=device)
+
+    # The core part of the update routine can be customized by overriding.
+    def update_core(self):
+        # When we pass one iterator and optimizer to StandardUpdater.__init__,
+        # they are automatically named 'main'.
+        train_iter = self.get_iterator('main')
+        optimizer = self.get_optimizer('main')
+        # Progress the dataset iterator for sentences at each iteration.
+        batch = train_iter.__next__()
+        x, t = convert.concat_examples(batch, device=self.device, padding=(0, -1))
+        # Concatenate the token IDs to matrices and send them to the device
+        # self.converter does this job
+        # (it is chainer.dataset.concat_examples by default)
+        xp = chainer.backends.cuda.get_array_module(x)
+        loss = 0
+        count = 0
+        state = None
+        batch_size, sequence_length = x.shape
+        for i in six.moves.range(sequence_length):
+            # Compute the loss at this time step and accumulate it
+            state, loss_batch = optimizer.target(state, chainer.Variable(x[:, i]),
+                                                 chainer.Variable(t[:, i]))
+            non_zeros = xp.count_nonzero(x[:, i])
+            loss += loss_batch * non_zeros
+            count += int(non_zeros)
+
+        reporter.report({'loss': float(loss.data)}, optimizer.target)
+        reporter.report({'count': count}, optimizer.target)
+        # update
+        loss /= batch_size  # normalized by batch size
+        optimizer.target.cleargrads()  # Clear the parameter gradients
+        loss.backward()  # Backprop
+        loss.unchain_backward()  # Truncate the graph
+        optimizer.update()  # Update the parameters
+
+
+class LMEvaluator(extensions.Evaluator):
```

Comment:
Same comment with the BPTTUpdater. We may use extensions.Evaluator instead of making a custom evaluator.

Reply:
Maybe possible to use the default evaluator. but it's better to do it in a different PR.